### ISSUE TITLE
Question 10

### ISSUE THREAD

**Issue by robindurkowitzer** [cmm-21/a6#24]
Instead of using many formulas I now tried to penalize the difference to an optimal velocity and position directly, which should give me the correct solution. However my Spaceship still does not stay in perfect orbit. 
I am not sure how to fix this issue. Is it a numerical issue? or where could it come from?

**Comment by MiguelZamoraM** [comment]
Well the fact that you reach a nominal velocity doesn't really guarantee that you will have a perfect orbit after you cut the thrusters. For that you might also need to simultaneously reach also a target acceleration and even then you might need to also account for some inaccuracies of the sim. The point here is more about assuming that the simulation is right and then use trajectory optimization tools to accomplish our goals.   

Take a look at https://github.com/cmm-21/a6/issues/14#issuecomment-845947773. I hope that helps.

**Comment by robindurkowitzer** [comment]
hmm, okay
What confuses me in that answer is that the objective on the state already depends on the control input itself, since the state depends on the control input.
How would we differently make the Objective function depend on the control input? Could you maybe elaborte on that a little more?

**Comment by MiguelZamoraM** [comment]
You are right. Indeed, since the state depends on the controls then an objective that depends on the state also depends _**implicitly**_ on the controls. 

However, when you add to your objective regularization terms like x.squaredNorm() and  u.squaredNorm(), you are actually specifying an _**explicit**_ dependency of your objective on states and the controls. (That way you are also making things easier to tune)

So the point is then how to specify _**explicitly**_ on the objective using the states and/or the controls, at different parts of the trajectory, the fact that we want the  ship in the future to remain in orbit after the thrusters are off. 

PS: The usage of the word explicit and implicit here has nothing to do with integration schemes.

**Comment by robindurkowitzer** [comment]
Alright that makes sense... 
The concept of the regularizers is also not copletely clear to me in this context. 
We introduced regularizers in Newton's method to make sure the Hessian is not too close to 0 and we are able to mix between Newton's method and gradient descent. 
Now in the challenge app, we use gradient descent. What influence do the regularizers have there? 
Are we then still able to fulfill the goal of having a desired end velocity and positon if we add those regularization terms?

**Comment by MiguelZamoraM** [comment]
Imagine you have an objective **TotObj** that is the result of a weighted sum of different subobjectives **sub_obj_i**.

In this context, by regularization I mean a subobjective that penalizes a quantity for being too big. This usually makes your optimization landscape to be a bit more well-bahaved.

Now in the challenge app you can come up with a subobjective that clearly specifies that you want your ship to stay in orbit. Let's call that **magic_orbit_subObj**.

Your total objective will be then something like: **TotObj** = c1* **magic_orbit_subObj** + c2 * **x.squaredNorm()** + c3 * **u.squaredNorm()**  + other stuff.

Now you can set the scalar coefficient c1 to a larger value, meaning that that is the sub-objective you care the most, and set coefficients c2 and c3 to smaller values to prevent the optimization from going to regions where the state or the controls are just too large. 

Notice also that by making the weight c1 larger, the  **magic_orbit_subObj** will dominate over the other sub-objectives, allowing you to fulfill the goal of making your ship stay in orbit. 

PS: The concept of large and small is completely relative, c1 could be 1 and c2 could be 1e-3, it's not like you always need 10e6.

**Comment by robindurkowitzer** [comment]
hmm alright, I guess I have to do some thinking about this magic objective then :D
 Thanks for the detailed explanation! 
 
 Edit: 
 I have been pondering about different imlementation of an orbiting objective, but I am very stuck in my line of thought. 
 Is there a different way of thinking about orbiting than meeting physical constraints? I am losing my mind at this 😅

**Comment by MiguelZamoraM** [comment]
As far as I understand you approach is based on reaching a certain position with a certain velocity that you estimate using your knowledge of physics. That can work but there are certainly other ways, or better said other objectives that can also be used to describe the goal of reaching the orbit and then staying in orbit even after thrusters are turned off.

**Comment by robindurkowitzer** [comment]
Finally got it! thanks!
